src/target/riscv/riscv-013.c

```diff
@@ -3396,329 +3396,567 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
-				 * attempted to read when we discovered that the target was
-				 * busy. */
-				if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
-				if (size > 4 && dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+	const uint32_t startup_command =
+		read_memory_progbuf_inner_startup_command(target);
+	if (execute_abstract_command(target, startup_command) != ERROR_OK)
+		return ERROR_FAIL;
 
-				/* See how far we got, clobbering dmi_data0. */
-				if (increment == 0) {
-					uint64_t counter;
-					result = register_read_direct(target, &counter, GDB_REGNO_A0);
-					next_index = counter;
-				} else {
-					uint64_t next_read_addr;
-					result = register_read_direct(target, &next_read_addr,
-												  GDB_REGNO_S0);
-					next_index = (next_read_addr - address) / increment;
-				}
-				if (result != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+	/* First read has just triggered. Result is in s1.
+	 * dm_data registers contain the previos value of s1 (garbage).
+	 */
+	if (dmi_write(target, DM_ABSTRACTAUTO,
+				set_field(0, DM_ABSTRACTAUTO_AUTOEXECDATA, 1)) != ERROR_OK)
+		return ERROR_FAIL;
 
-				uint64_t value64 = (((uint64_t)dmi_data1) << 32) | dmi_data0;
-				buf_set_u64(buffer + (next_index - 2) * size, 0, 8 * size, value64);
-				log_memory_access64(address + (next_index - 2) * size, value64, size, true);
+	/* Read garbage from dm_data0, which triggers another execution of the
+	 * program. Now dm_data contains the first good result (from s1),
+	 * and s1 the next memory value.
+	 */
+	if (dmi_read_exec(target, NULL, DM_DATA0) != ERROR_OK) {
+		dmi_write(target, DM_ABSTRACTAUTO, 0);
+		return ERROR_FAIL;
+	}
 
-				/* Restore the command, and execute it.
-				 * Now DM_DATA0 contains the next value just as it would if no
-				 * error had occurred. */
-				dmi_write_exec(target, DM_COMMAND, command, true);
-				next_index++;
+	uint32_t abstractcs;
 
-				dmi_write(target, DM_ABSTRACTAUTO,
-						1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+	if (wait_for_idle(target, &abstractcs))
+		return ERROR_FAIL;
 
-				ignore_last = 1;
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		break;
+	default:
+		/*TODO: Here one could attempt to try again on CMDERR_BUSY */
+		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
+				info->cmderr);
+		riscv013_clear_abstract_error(target);
+		dmi_write(target, DM_ABSTRACTAUTO, 0);
+		return ERROR_FAIL;
+	}
+	return ERROR_OK;
+}
 
-				break;
-			default:
-				LOG_DEBUG("error when reading memory, abstractcs=0x%08lx", (long)abstractcs);
-				riscv013_clear_abstract_error(target);
-				riscv_batch_free(batch);
-				result = ERROR_FAIL;
-				goto error;
-		}
+static uint32_t read_memory_progbuf_inner_fill_batch(struct riscv_batch *batch,
+		uint32_t count, uint32_t size);
```

Comment:
For this and other prototypes, is it possible to reorder the functions to eliminate the prototypes?

Since this change already rewrites all of it, there's not downside to moving the functions around.

Reply:
These prototypes are not needed in most cases (for now I only see one "loop" : `read_memory_progbuf_inner_loop_iteration` calls `read_memory_progbuf_inner_ensure_forward_progress`, which calls `read_memory_progbuf_inner_loop_iteration`).

I have added these to reorder the code in such a way that when the code is ridden from the top, the use of a function is seen before the definition. IMHO, this makes code easier to understand.

One example of such a case is `read_memory_progbuf_inner_startup`. When functions are reordered, this one appears at the top. But it does not make much sense on its own (without the context of `read_memory_progbuf_inner`). Furthermore, it would be separated from `read_memory_progbuf_inner` by a plethora of the likes of `read_memory_progbuf_inner_on_dmi_busy` which would not make sense either, so by the time a reader reaches `read_memory_progbuf_inner` they might get confused.

However, I understand that this is not consistent with the codebase. So, if you don't like it and don't find my reasoning appealing enough, I will gladly reorder these functions.

Reply:
Whether it's easier to see users or callers first depends on the code and the reader. Usually my preference is to eliminate prototypes because they're unnecessary duplication.

When reading the code I assume people mostly rely on the tools their editors give them to navigate between functions. (Unfortunately in github reviews this is not easy.)